executor/common_zlib.h

```diff
@@ -0,0 +1,530 @@
+// Copyright 2022 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+// This code is derived from puff.{c,h}, found in the zlib development. The
+// original files come with the following copyright notice:
+
+// Copyright (C) 2002-2013 Mark Adler, all rights reserved
+// version 2.3, 21 Jan 2013
+// This software is provided 'as-is', without any express or implied
+// warranty.  In no event will the author be held liable for any damages
+// arising from the use of this software.
+// Permission is granted to anyone to use this software for any purpose,
+// including commercial applications, and to alter it and redistribute it
+// freely, subject to the following restrictions:
+// 1. The origin of this software must not be misrepresented; you must not
+// 		claim that you wrote the original software. If you use this software
+// 		in a product, an acknowledgment in the product documentation would be
+// 		appreciated but is not required.
+// 2. Altered source versions must be plainly marked as such, and must not be
+// 		misrepresented as being the original software.
+// 3. This notice may not be removed or altered from any source distribution.
+// Mark Adler    madler@alumni.caltech.edu
+
+// All dynamically allocated memory comes from the stack.  The stack required
+// is less than 2K bytes.  This code is compatible with 16-bit int's and
+// assumes that long's are at least 32 bits.  puff.c uses the short data type,
+// assumed to be 16 bits, for arrays in order to conserve memory.  The code
+// works whether integers are stored big endian or little endian.
+
+#include <setjmp.h> // for setjmp(), longjmp(), and jmp_buf
+
+// Maximums for allocations and loops.
+#define MAXBITS 15 // maximum bits in a code
+#define MAXLCODES 286 // maximum number of literal/length codes
+#define MAXDCODES 30 // maximum number of distance codes
+#define MAXCODES (MAXLCODES + MAXDCODES) // maximum codes lengths to read
+#define FIXLCODES 288 // number of fixed literal/length codes
+
+struct state {
+	// output state
+	unsigned char* out; // output buffer
+	unsigned long outlen; // available space at out
+	unsigned long outcnt; // bytes written to out so far
+
+	// input state
+	const unsigned char* in; // input buffer
+	unsigned long inlen; // available input at in
+	unsigned long incnt; // bytes read so far
+	int bitbuf; // bit buffer
+	int bitcnt; // number of bits in bit buffer
+
+	// input limit error return state for bits() and decode()
+	jmp_buf env;
+};
+
+// Return need bits from the input stream.  This always leaves less than
+// eight bits in the buffer.  bits() works properly for need == 0.
+static int bits(struct state* s, int need)
```

Comment:
Because this is a header, all these rather generic function names, like 'bits', 'stored', 'decode', etc. as well as 'struct state' have a high chance of name clashes.

It seems to work today, but it might make sense to prefix these with 'puff_' or something else suitable.